### ISSUE TITLE
[Dart-dio-next] nullable fields sent when using multipart forms

### ISSUE THREAD

**Issue by jaumard** [OpenAPITools/openapi-generator#10295]
This is what is generated today with last version: 

```
Future<Response<User>> saveProfile({ 
    String? email,
    String? firstName,
    String? lang,
    String? lastName,
    String? mobile,
    String? password,
    MultipartFile? avatar,
    CancelToken? cancelToken,
    Map<String, dynamic>? headers,
    Map<String, dynamic>? extra,
    ValidateStatus? validateStatus,
    ProgressCallback? onSendProgress,
    ProgressCallback? onReceiveProgress,
  }) 
....

_bodyData = FormData.fromMap(<String, dynamic>{
        r'email': encodeFormParameter(_serializers, email, const FullType(String)),
        r'firstName': encodeFormParameter(_serializers, firstName, const FullType(String)),
        r'lang': encodeFormParameter(_serializers, lang, const FullType(String)),
        r'lastName': encodeFormParameter(_serializers, lastName, const FullType(String)),
        r'mobile': encodeFormParameter(_serializers, mobile, const FullType(String)),
        r'password': encodeFormParameter(_serializers, password, const FullType(String)),
        r'avatar': avatar,
      });
```

Problem with that is that null fields are still sent to the payload, and even worse  `encodeFormParameter` is forcing null field to be empty strings. 

Basically here I don't want an empty string sent as a new password ^^, so I'm getting 400 all the time because password is not correct.   

I guess an easy fix would be to wrap then into a if like this:
_bodyData = FormData.fromMap(<String, dynamic>{
        if(email != null)  r'email': encodeFormParameter(_serializers, email, const FullType(String)),
        if(firstName != null)  r'firstName': encodeFormParameter(_serializers, firstName, const FullType(String)),
        if(lang != null) r'lang': encodeFormParameter(_serializers, lang, const FullType(String)),
        if(lastName != null) r'lastName': encodeFormParameter(_serializers, lastName, const FullType(String)),
        if(mobile != null) r'mobile': encodeFormParameter(_serializers, mobile, const FullType(String)),
        if(password != null) r'password': encodeFormParameter(_serializers, password, const FullType(String)),
        if(avatar != null) r'avatar': avatar,
      });
```

**Comment by kuhnroyal** [comment]
Hmm, I don't think it's that easy but should be solvable.
We probably need to check the `required` and `isNullable` flags as well.

**Comment by jaumard** [comment]
Yes wrapping if should be based on those two flags for sure. I saying it's "easy" because we already have all the info in the template to make it work, so no heavy development in java or something ^^

**Comment by kuhnroyal** [comment]
So the template actually contains `{{^required}}{{^isNullable}}if ({{{paramName}}} != null) {{/isNullable}}{{/required}}`.
What are you parameters configured like?

**Comment by jaumard** [comment]
Wow let me check maybe I did shit on my side like a noob ^^

**Comment by jaumard** [comment]
So this is what I have: 

```
/api/v1/users/me:
    patch:
      operationId: saveProfile
      description: ""
      requestBody:
        content:
          multipart/form-data:
            schema:
              "$ref": "#/components/schemas/UserUpdate"
UserUpdate:
      type: object
      properties:
        email:
          type: string
          format: email
          nullable: true
        firstName:
          type: string
          nullable: true
        lang:
          type: string
          nullable: true
        lastName:
          type: string
          nullable: true
        mobile:
          type: string
          nullable: true
        password:
          type: string
          nullable: true
        avatar:
          type: string
          format: binary
          nullable: true
```

Look correct to me.

**Comment by kuhnroyal** [comment]
Make it `nullable: false`?

It is not clear from the spec what to do here:
> Note: nullable is not the same as an optional parameter or an empty-valued parameter. nullable means the parameter value can be null. Specific implementations may choose to map an absent or empty-valued parameter to null, but strictly speaking these are not the same thing.

So we have to send `password=`, not sure about form data and how `null` is handled vs an empty string.

**Comment by jaumard** [comment]
Spec says implementation can choose ^^ 
To me sending empty strings on a PATCH request totally breaks the fact that it's a PATCH you that you can update only what is sent.  

On my side that's not possible to put `nullable: false` and ask my user to enter his password each time he want to update his info. 

Also for example the avartar field will not be sent when it's null, so for same request we have two different behavior. Doesn't feel right.

**Comment by kuhnroyal** [comment]
How is `null` vs `''` encoded in the form data?

`nullable: false` is not the same as `required`.

**Comment by jaumard** [comment]
Currently null values are not sent, not sure if it's dio doing that or not, but that's what append. And `encodeFormParameter` is forcing empty string when the value is null, maybe the fix is just to return null if the given value is null and not forcing it to empty string.

**Comment by kuhnroyal** [comment]
We can do that but you should also try `nullable: false`, that should work as well.

**Comment by jaumard** [comment]
Just tried the `nullable: false, ` just not to be stubborn and dumb lol looks like you're right it works! But then I have trouble understanding how it works between required and nullable. 

To me in my mind it should work like this: 
nullable = false => generated method must have non nullable field
nullable = true => generated method must have nullable field 
required = true => field is part of the payload (null or not depending on nullable)
required = false => field not part of the payload if null

And I think that's how it work when dealing with JSON payload. 
Here I found that it works weirdly because whatever I put for nullable my dart method accepts nullable. Doesn't feel right to me. 
What do you think?

**Comment by kuhnroyal** [comment]
Well it doesn't work exactly this way in Dart.
The combination `nullable: false` and `required: false` in OAS3 still needs to be a nullable field in Dart, otherwise you either need to always specify the parameter or it needs a default value.

And that is exactly the case you are in.
You don't require the field and you don't want to send `null`. 

> required = false => field not part of the payload if null

I think this part makes sense, the rest we basically already do.

**Comment by kuhnroyal** [comment]
So for Dart:

`nullable: true` & `required: true` = `required String? foo`
`nullable: false` & `required: true` = `required String foo`
`nullable: true` & `required: false` = `String? foo`
`nullable: false` & `required: false` = `String? foo` - This is the special case you are running into that can not mirror the OAS spec. `String foo` is not allowed, only `required String foo` or `String foo = 'bar'`, both of those don't work for us.

**Comment by jaumard** [comment]
Yes 
But maybe `encodeFormParameter ` should return null if value is null no? In case of required:true & nullable: true, cause currently, it forces empty strings.

**Comment by kuhnroyal** [comment]
Yea, already did that locally.

**Comment by albe-jj** [comment]
Hi, if anyone is having this issue form a FastAPI generated openapi spces, the solution is to add `nullable=True` to your field or query

see for instance

```python
get_scopes(*, name: Optional[str] = Query(None, max_length=100, nullable=True),
```

this way in the openapi.json it will be declared as `nullable:True`.

Note more info on why `nullable=True` is required is described in details here https://github.com/fastapi/fastapi/issues/2924

**Comment by jonasN5** [comment]
`required = false => field not part of the payload if null` this doesn't seem like it's been implemented, has it? This is a major blocker for me and I have to patch it locally. Always passing the parameter in the payload and on top casting to an empty string causes big headaches on the backend.
I'm wondering how I can reset a field though? Say a field is of type int is nullable on the backend. Now I want to remove an existing value, which means I have to pass null to the backend. How does that currently work?

I don't think it's currently possible. The method describe above should instead be:

```
Future<Response<User>> saveProfile({Map<String, dynamic>? body,
    CancelToken? cancelToken,
    Map<String, dynamic>? headers,
    Map<String, dynamic>? extra,
    ValidateStatus? validateStatus,
    ProgressCallback? onSendProgress,
    ProgressCallback? onReceiveProgress,
  }) 
```

And then we look for each key in body and if the key is present, serialize it properly.

**Comment by Christiaan-EPSza** [comment]
I'm also hitting this issue — required: false query parameters are still being included in the request even when their value is null.

The root cause, as identified above, is the template condition:

The {{^isNullable}} guard means the null check is skipped for nullable: true params. But since required: false params must be T? in Dart regardless of nullability, the null guard should always be emitted for non-required params — the isNullable distinction is irrelevant here.

The proposed fix of changing this to {{^required}}if ({{{paramName}}} != null) {{/required}} is correct and would resolve it.

As a workaround we're currently stripping nullable: true from query parameters in the spec before generation, which is not ideal.

@wing328  — first off, thank you for all the work you put into this project, it's genuinely appreciated. When you get a moment, would you be able to share any thoughts on this? Even just a pointer on the right direction to get a fix moving would be really helpful for the community.
